### PR TITLE
[SPARK-44022][BUILD] Enforce max bytecode version on Maven dependencies

### PR DESCRIPTION
**Pull request by bowenliang123**
### What changes were proposed in this pull request?

- to enforce Java's max bytecode version to maven dependencies, by using enforceBytecodeVersion enforcer rule (https://www.mojohaus.org/extra-enforcer-rules/enforceBytecodeVersion.html)
- exclude `org.threeten:threeten-extra` in enforcer rule as its package-info.class requiring bytecode version 53 but no side effects on other classes
 
### Why are the changes needed?

- to avoiding introduction of dependencies requiring higher Java version

### CODE REVIEW

**Comment by LuciferYang** [comment]
So we can't use `org.threeten:threeten-extra:jar:1.7.1` ?

When I run `build/mvn clean install -Phadoop-3 -Phadoop-cloud -Pmesos -Pyarn -Pkinesis-asl -Phive-thriftserver -Pspark-ganglia-lgpl -Pkubernetes -Phive -DskipTests`

There are following compile errors:

```
[INFO] --- maven-enforcer-plugin:3.3.0:enforce (enforce-versions) @ spark-sql_2.12 ---
[INFO] Rule 2: org.apache.maven.enforcer.rules.dependency.BannedDependencies passed
[INFO] Adding ignore: module-info
[INFO] Restricted to JDK 8 yet org.threeten:threeten-extra:jar:1.7.1:compile contains org/threeten/extra/chrono/package-info.class targeted to JDK 9
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Spark Project Parent POM 3.5.0-SNAPSHOT:
[INFO] 
[INFO] Spark Project Parent POM ........................... SUCCESS [  8.066 s]
[INFO] Spark Project Tags ................................. SUCCESS [  8.060 s]
[INFO] Spark Project Sketch ............................... SUCCESS [  8.749 s]
[INFO] Spark Project Local DB ............................. SUCCESS [ 10.234 s]
[INFO] Spark Project Networking ........................... SUCCESS [ 14.018 s]
[INFO] Spark Project Shuffle Streaming Service ............ SUCCESS [ 12.281 s]
[INFO] Spark Project Unsafe ............................... SUCCESS [ 11.616 s]
[INFO] Spark Project Common Utils ......................... SUCCESS [  9.951 s]
[INFO] Spark Project Launcher ............................. SUCCESS [  8.825 s]
[INFO] Spark Project Core ................................. SUCCESS [02:23 min]
[INFO] Spark Project ML Local Library ..................... SUCCESS [ 25.190 s]
[INFO] Spark Project GraphX ............................... SUCCESS [ 28.595 s]
[INFO] Spark Project Streaming ............................ SUCCESS [ 39.839 s]
[INFO] Spark Project Catalyst ............................. SUCCESS [02:42 min]
[INFO] Spark Project SQL .................................. FAILURE [  0.769 s]
[INFO] Spark Project ML Library ........................... SKIPPED
[INFO] Spark Project Tools ................................ SKIPPED
[INFO] Spark Project SQL API .............................. SKIPPED
[INFO] Spark Project Hive ................................. SKIPPED
[INFO] Spark Project REPL ................................. SKIPPED
[INFO] Spark Project YARN Shuffle Service ................. SKIPPED
[INFO] Spark Project YARN ................................. SKIPPED
[INFO] Spark Project Mesos ................................ SKIPPED
[INFO] Spark Project Kubernetes ........................... SKIPPED
[INFO] Spark Project Hive Thrift Server ................... SKIPPED
[INFO] Spark Ganglia Integration .......................... SKIPPED
[INFO] Spark Project Hadoop Cloud Integration ............. SKIPPED
[INFO] Spark Project Assembly ............................. SKIPPED
[INFO] Kafka 0.10+ Token Provider for Streaming ........... SKIPPED
[INFO] Spark Integration for Kafka 0.10 ................... SKIPPED
[INFO] Kafka 0.10+ Source for Structured Streaming ........ SKIPPED
[INFO] Spark Kinesis Integration .......................... SKIPPED
[INFO] Spark Project Examples ............................. SKIPPED
[INFO] Spark Integration for Kafka 0.10 Assembly .......... SKIPPED
[INFO] Spark Avro ......................................... SKIPPED
[INFO] Spark Project Connect Common ....................... SKIPPED
[INFO] Spark Protobuf ..................................... SKIPPED
[INFO] Spark Project Connect Server ....................... SKIPPED
[INFO] Spark Project Connect Client ....................... SKIPPED
[INFO] Spark Project Kinesis Assembly ..................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  08:13 min
[INFO] Finished at: 2023-06-12T11:09:01+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.3.0:enforce (enforce-versions) on project spark-sql_2.12: 
[ERROR] Rule 3: org.codehaus.mojo.extraenforcer.dependencies.EnforceBytecodeVersion failed with message:
[ERROR] Found Banned Dependency: org.threeten:threeten-extra:jar:1.7.1
[ERROR] Use 'mvn dependency:tree' to locate the source of the banned dependencies.

```

**Comment by bowenliang123** [comment]
Yes, it's an known conflicts and that's why we pinned threeten to 1.7.0 in Kyuubi. https://github.com/apache/kyuubi/pull/4292

**Comment by bowenliang123** [comment]
But I'm still wondering how threeten 1.7.1 violates the bytecode version rule, as I don't see conflicts in randomly chosen classes file from threeten-extra 1.7.1 jar.

**Comment by LuciferYang** [comment]
@bowenliang123 In what scenarios will it trigger what problems? So far, GA has passed. 

Does SBT have a corresponding plugin? Or add a maven build with Java 8 task for GA? 

According to this rule, can't we also use Jackson 2.15? But all UTs have passed, does this really have a practical impact?

**Comment by LuciferYang** [comment]
cc @dongjoon-hyun FYI due to `org.threeten:threeten-extra:jar:1.7.1` is  a cascading dependency of `org.apache.orc:orc-core:jar:shaded-protobuf:1.8.3`

**Comment by bowenliang123** [comment]
It's not triggering any problems so far.
This PR aims to provide improvement for a checked rather than an assumed baseline for bytecode version alignment.
There are similar enforcer plugins for Gradle and Maven but not sbt.
With this rule, it's good to use Jackson 2.15 (as we do in Kyuubi now).

So far, this PR intends to raise a discussion in bytecode alignment and to recognize the threaten dependency triggering the rule. It will be kept as a draft until we have a further agreement on this.

**Comment by LuciferYang** [comment]
So should we add a maven build GA task using Java 8? Otherwise, nothing is checked before merging, which will become an afterthought rule ...

File: .github/workflows/build_and_test.yml
Comment (LuciferYang) [inline .github/workflows/build_and_test.yml:730]: `-Denforcer.skip=true` when using Java 8?

Comment (bowenliang123) [reply .github/workflows/build_and_test.yml:730]: I thought you asked for a build on Java 8 to see what happens. Skipping enforcer rule for Java 8 to check whether it causes runtime exceptions when loading threeten jar claseses.

Comment (LuciferYang) [reply .github/workflows/build_and_test.yml:730]: No, what I mean is we should do this check when user submitting code, especially for Java 8, otherwise the addition of this plugin will not have its intended effect, so I think `enforcer.skip` should always be false

Comment (bowenliang123) [reply .github/workflows/build_and_test.yml:730]: Are you looking for this? log: https://github.com/bowenliang123/spark/actions/runs/5239841838/jobs/9460075811#step:11:69

It's checked along with `mvn validate` in CI builds.

**Comment by bowenliang123** [comment]
After checking the major version of every class files from `threeten-extra` jar file 1.7.1 (also latest 1.7.2), it turns out three package info classes violates the bytecode enforcer rule.
```
threeten-extra-1.7.1/org/threeten/extra/scale/package-info.class, major version:53
threeten-extra-1.7.1/org/threeten/extra/chrono/package-info.class, major version:53
threeten-extra-1.7.1/org/threeten/extra/package-info.class, major version:53
```


Manual checks on class file with ` javap -v threeten-extra-1.7.1/org/threeten/extra/package-info.class`,
```
  Last modified 2022-8-2; size 124 bytes
  MD5 checksum 6e7e78b5c0db672c93335f76c78dc092
  Compiled from "package-info.java"
interface org.threeten.extra.package-info
  minor version: 0
  major version: 53
...
SourceFile: "package-info.java"
```

File: .github/workflows/build_and_test.yml
Comment (LuciferYang) [reply .github/workflows/build_and_test.yml:730]: Do you mean `./dev/test-dependencies.sh` already covered this inspection? That's great, sorry for misleading, then I think we don't need to add any extra Maven builds anymore.

Comment (bowenliang123) [reply .github/workflows/build_and_test.yml:730]: Yes, it does as you expected. Thanks.

Comment (LuciferYang) [reply .github/workflows/build_and_test.yml:730]: Thanks for your explain

Comment (LuciferYang) [reply .github/workflows/build_and_test.yml:730]: Let's revert this change :)

Comment (bowenliang123) [reply .github/workflows/build_and_test.yml:730]: Do you think I should unset Java 8  from the matrix, and skip maven builds on it?

Comment (bowenliang123) [reply .github/workflows/build_and_test.yml:730]: OK, sure will do it. And I'm curious to see a full build run on Java 8. And revert the change after that.

**Comment by LuciferYang** [comment]
So should we explicitly add a more appropriate `org.threeten:threeten-extra` dependency in this pr?

**Comment by bowenliang123** [comment]
> So should we explicitly add a more appropriate `org.threeten:threeten-extra` dependency in this pr?

**Comment by bowenliang123** [comment]
> So should we explicitly add a more appropriate `org.threeten:threeten-extra` dependency in this pr?

There's no appropriate version of `threeten-extra` for solving the violation to this rule, as 1.7.1 (used by orc in spark-sql module now) and 1.7.2 (latest available release version)  has the same problem. Also, `threeten` only requires Java 8 . `ThreeTen-Extra requires Java SE 8 or later and has no dependencies.` (quote from https://github.com/ThreeTen/threeten-extra)

So the only workaround for this is adding an exclusion for `threeten-extra` in `enforceBytecodeVersion` rule.
```
<enforceBytecodeVersion>
    <maxJdkVersion>${java.version}</maxJdkVersion>
    <ignoredScopes>test</ignoredScopes>
    <excludes>
      <exclude>org.threeten:threeten-extra</exclude>
    </excludes>
</enforceBytecodeVersion>
```

File: .github/workflows/build_and_test.yml
Comment (bowenliang123) [reply .github/workflows/build_and_test.yml:730]: Success builds with Maven on Java 8 with enforcer skipped.
https://github.com/bowenliang123/spark/actions/runs/5239841838/jobs/9460063427#step:7:7

**Comment by pan3793** [comment]
> ... adding an exclusion for `threeten-extra` in `enforceBytecodeVersion` rule.

it's fine, because it's not a real problem.

https://github.com/apache/spark/pull/37787#issuecomment-1424173443

> I checked the files inside threeten-extra-1.7.1.jar, and only module-info.class has the issue, other classes can be loaded by JDK8 correctly. Thus IMO it's not a big problem.

File: pom.xml
Comment (LuciferYang) [inline pom.xml:2788]: with this change, there still compile failed as follows:

```
[INFO] Spark Project Connect Client ....................... FAILURE [  0.839 s]
[INFO] Spark Project Kinesis Assembly ..................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  22:18 min
[INFO] Finished at: 2023-06-12T14:28:19+08:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.3.0:enforce (enforce-versions) on project spark-connect-client-jvm_2.12: 
[ERROR] Rule 3: org.codehaus.mojo.extraenforcer.dependencies.EnforceBytecodeVersion failed with message:
[ERROR] Found Banned Dependency: io.get-coursier:interface:jar:1.0.11
[ERROR] Use 'mvn dependency:tree' to locate the source of the banned dependencies.
```

Comment (bowenliang123) [reply pom.xml:2789]:
```
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- dependency:3.5.0:tree (default-cli) @ spark-connect-client-jvm_2.12 ---
[INFO] org.apache.spark:spark-connect-client-jvm_2.12:jar:3.5.0-SNAPSHOT
[INFO] \- com.lihaoyi:ammonite_2.12.17:jar:2.5.8:provided
[INFO]    \- com.lihaoyi:ammonite-interp-api_2.12.17:jar:2.5.8:provided
[INFO]       \- io.get-coursier:interface:jar:1.0.11:provided
```
`io.get-coursier:interface:jar` is used in provided scope.

Solved this problem by skipping `provided` scope for enforceBytecodeVersion rule.

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for making a PR, @bowenliang123 . Thank you for pinging me, @LuciferYang .

Yes, Apache ORC community found this before, but we also didn't find any functional issue because it's a kind of metadata file.

I'm the release manager of Apache ORC 1.9.0 and Apache ORC community will release it soon. I guess we may resolve this exceptional case from Apache ORC community first.

Let me take a look at the plugin more.

File: pom.xml
Comment (bowenliang123) [reply pom.xml:2789]: OK, a subtask [SPARK-44032](https://issues.apache.org/jira/browse/SPARK-44032) added for this, and updated the todo comment as well.

**Review by LuciferYang — APPROVED** [review]

**Comment by bowenliang123** [comment]
Thanks for the reply. And the problem in `threeten-extra` has been fixed in https://github.com/ThreeTen/threeten-extra/pull/260 by @cxzl25, and should be available in their next release. Before that, I think we should keep this exclusion for `threeten-extra` in the bytecode version rule.

@dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
Thank you for the additional info. Let's hold on this PR then.

**Comment by bowenliang123** [comment]
Alright, this PR is ready. We could have this PR merged first to prevent other possible bytecode version violations from other bumping or introduction of maven dependencies. I don't think `threeten-extra` will have the next release with the fix soon, as none of any other changes/commits except to it.

**Review by dongjoon-hyun — DISMISSED** [review]
Thank you for making this PR and validating the dependency, @bowenliang123 and @LuciferYang .

1. Technically, this PR is only for testing and preventing newly added dependency and we don't change any dependency here.

2.  Given that Apache Spark community has a policy which backports only bug fixes, this PR will not be backported to branch-3.4. So, there is no urgency for Apache Spark 3.4.1.

3. For Apache Spark 3.5.0 target,  we have still over one month until July 16th for Apache Spark 3.5.0 Feature Freeze. Also, we have no urgency for this PR.
  - https://spark.apache.org/versioning-policy.html

4. Lastly, `Target Version` field is reserved for Apache Spark committer. Please remove that field to avoid future confusions.
  - https://spark.apache.org/contributing.html
> Do not set the following fields:
> Target Version. This is assigned by committers to indicate a PR has been accepted for possible fix by the target version.

Screenshot 2023-06-12 at 8 45 17 PM

**Comment by bowenliang123** [comment]
I understand and well agree with your detailed explanations and suggestions. 
Already updated Jira Issues [SPARK-44022](https://issues.apache.org/jira/browse/SPARK-44022) and the todo subtask  [SPARK-44032](https://issues.apache.org/jira/browse/SPARK-44032) with target version unset.
It's my first initial PR to Spark and sorry for the inconvenience it brought to you guys.

**Comment by dongjoon-hyun** [comment]
Thank you so much and never mind. My initial comment wasn't enough to give the full context. It's my bad.

**Review by dongjoon-hyun — APPROVED** [review]
Sorry for making you wait for a long time. After checking the situation more, I also agree with you now.

+1, LGTM for Apache Spark 3.5.0.

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 3.5.0.

Welcome to the Apache Spark community and congratulation for your first commit, @bowenliang123 .
I added you to the Apache Spark contributor group and assigned SPARK-44022 to you. Thank you again!

**Comment by bowenliang123** [comment]
Thanks for the guidance and help from @LuciferYang and @dongjoon-hyun. Although it's just a minor supportive contribution to Spark, it's still my pleasure to have my first contribution to this great project.
And I would like to thank my colleague @jeanlyn at Guangfa Securities here, who contributed serval commits to Spark years ago. He and his contribution inspire me that we as users and participants in the big data ecosystem, is able to contribute to top-level projects. As for @pan3793 , here's my thanks all the way in this PR, in Spark and in Kyuubi.

**Pull request closed without merging**
